### PR TITLE
The first implementation of `import ... from`

### LINKED ISSUE #8: Support for multiple files
As outlined in [ADR001](https://github.com/informalsystems/tnt/blob/main/doc/adr001-transpiler-architecture.md#33-resolve), we have to implement the task "resolve", which would resolve module imports.

### PR DESCRIPTION
**Pull request by konnov**
Closes #8. This is the first batch of support for `import .. from '<filename>'`. I had to keep balance between hacking it quickly and making it not too hacky. ~This PR only has happy path tests.~ The construct itself is working now. ~But we would need a follow up PR that adds tests for cyclic imports and other error processing~.

~Now it is working with `parse` but will probably fail with `run`, `test`, and REPL.~

It is working with `parse`, `run`, `test`, and vscode. REPL requires some googling #829.

Follow-up issues: #829, #830, #831, #832.

- [x] Tests added for any new code
- [x] Entries added to the respective `CHANGELOG.md` for any new functionality
- [x] Feature table on [`README.md`](../README.md#roadmap) updated for any listed functionality

### CODE REVIEW

File: quint/src/quintParserFrontend.ts
Comment (bugarela) [inline quint/src/quintParserFrontend.ts:141]: I got a sense of what this is doing, but how/why it works is not completely clear to me. It's 85 lines of imperative code, and I find it hard to follow. I'd much rather see this broken down into functions so I can reason about cycle detection, sorting and recursion separately.

I don't want to block this since we are just trying to get this working to unblock other work, but I do want to note the problem.

File: quint/src/sourceResolver.ts
Comment (bugarela) [inline quint/src/sourceResolver.ts:72]: What is `stem` here? I don't remember seeing it before and couldn't find the meaning with a couple of google searches.

File: quint/src/sourceResolver.ts
Comment (bugarela) [inline quint/src/sourceResolver.ts:112]: To avoid this duplication, we could have the `replacer` be an optional parameter to `fileSourceResolver` (with the identity function as its default value) and apply the `replacer` at the original `load call`, getting rid of `fileSourceResolverForTests` entirely.

File: quint/src/sourceResolver.ts
Comment (bugarela) [inline quint/src/sourceResolver.ts:166]:
```suggestion
}

```

File: quint/test/sourceResolver.test.ts
Comment (bugarela) [inline quint/test/sourceResolver.test.ts:99]:
```suggestion
})

```

**Review by bugarela — APPROVED** [review]
That's a lot of things you got working in a short amount of time :rocket:. I have some comments, but don't want to block this.

Also, could you add `from` to the list of keywords in our editor plugins?

File: quint/src/sourceResolver.ts
Comment (konnov) [reply quint/src/sourceResolver.ts:72]: It is the directory relative to the path of the importing file. I wanted to call it `basepath` but that is also confusing.

File: quint/src/quintParserFrontend.ts
Comment (konnov) [reply quint/src/quintParserFrontend.ts:141]: Yeah, it's not always easy to express the same thing in functional code. This code is using the [Worklist algorithm](https://en.wikipedia.org/wiki/Reaching_definition#Worklist_algorithm). The challenge here is to also detect which modules should go first, hence, it assigns ranks to modules. I think I could rewrite it into functional code, but it would be even less obvious than the imperative code.

File: quint/src/cliCommands.ts
Comment (shonfeder) [inline quint/src/cliCommands.ts:17]: can we give the parse phases descriptive names? `1b` doesn't mean anything. how about `loadFiles` or something?

File: examples/cosmos/ics23/ics23.qnt
Comment (shonfeder) [inline examples/cosmos/ics23/ics23.qnt:212]: This shouldn't be necessary iiuc? `from` seems to only be parsed out of import statements?

File: quint/src/quintParserFrontend.ts
Comment (shonfeder) [inline quint/src/quintParserFrontend.ts:136]:
```suggestion
 *  - parse the modules that are referenced by each path,
```

File: quint/src/quintParserFrontend.ts
Comment (shonfeder) [reply quint/src/quintParserFrontend.ts:141]: If you've referenced that algo in your implementation, would help to add a comment citing that.

File: quint/src/quintParserFrontend.ts
Comment (shonfeder) [inline quint/src/quintParserFrontend.ts:152]: It doesn't look like you reassign this variable. iiuc, this should be a const. it is a constant reference to a mutable structure: not a mutable reference. afaik, you'll still be able to push into it.

```suggestion
  const worklist: [QuintModule, SourceLookupPath[]][] =
```

**Review by shonfeder — CHANGES_REQUESTED** [review]
Quick turnaround here! :muscle: 

I'm part way thru, but would like to have the first baths of questions/suggestions addressed before merging.

File: quint/src/cliCommands.ts
Comment (bugarela) [inline quint/src/cliCommands.ts:189]: Could we instantiate this resolver inside `parsePhase1b`? It would be great if the top-level calls are as simple as possible since we will have to replicate this in the language server.

File: examples/cosmos/ics23/ics23.qnt
Comment (konnov) [reply examples/cosmos/ics23/ics23.qnt:212]: The ANTLR grammar introduces  a new token called 'from'. Once we have it, `from` should be a reserved keyword.

**Comment by shonfeder** [comment]
Posted my last batch of pending comments and removed my change request.

**Comment by shonfeder** [comment]
Not to block, but to record my uncertainty for the record, and maybe so I'd learn the motivation for the approach here:

The general thing I don't understand about this PR's approach is why we cannot approach the problem here in a shape a bit more like this:

- Do a pass to recursively phase1 parse all files imported by file reference
- in the process, rewrite relative file imports paths to absolute paths
- Extend the phase1 result with a mapping `absFilePath -> IR`
- Use this structure for dependency ordering (this part likely using the imperative algorithm, but only for benign state, without need for I/O. This is really just a special case of topological sorting iiuc?).
- Pass this ordered collection of modules on for module resolution.

The main benefits of this approach to my mind are

- Easier to reason about the separate operations.
- Don't need to mix file I/O into the relatively subtle dep resolution code

File: quint/src/quintParserFrontend.ts
Comment (shonfeder) [reply quint/src/quintParserFrontend.ts:149]: but why? Anything that has a reference to the map thru this variable will have it's ref mutated anyhow, so why mutate the reference and create a copy of a mutable structure instead of just mutating the mutable structure? Does something at some point keep a reference to the previous map?

Comment (konnov) [reply quint/src/quintParserFrontend.ts:149]: I do not want to mutate the output of phase1 directly. Phase 1 is called multiple times, once per file. So instead of collecting the source information in the source map of the main file, I am making a copy here.

Comment (shonfeder) [reply quint/src/quintParserFrontend.ts:149]: sorry if i'm being dense, but i'm afraid I don't follow: you are already copying the Phase1 output into a map via the `new Map` call, no? Each time you get a new Phase1 result you are adding it to this map. Whether you mutate this map or mutate the ref to add a new map, you are not touching the phase1 result in directly either way, are you?

Comment (shonfeder) [reply quint/src/quintParserFrontend.ts:149]: To be clear, this isn't a blocking objection, but i just don't see the benefit to constantly creating copies if you are mutating the state anyhow.

File: quint/src/sourceResolver.ts
Comment (konnov) [reply quint/src/sourceResolver.ts:45]: I've introduced this abstraction, as it has been quite an issue in SANY. We do not have to deal with filename, in order to resolve relative paths. In case of tests, we could use a map of strings. For a web interface, we could use a map of names to URIs. It's true that the main use case is filenames right now.

File: quint/src/sourceResolver.ts
Comment (konnov) [reply quint/src/sourceResolver.ts:88]: toSourceName produces a platform-independent path, since it is using `posix.join`, the paths will look the same on Linux and Windows. I actually had this issue with tests.

File: quint/src/cliCommands.ts
Comment (konnov) [reply quint/src/cliCommands.ts]: Yeah, I have renamed all phases

File: quint/src/cliCommands.ts
Comment (konnov) [reply quint/src/cliCommands.ts:189]: I agree with the intention. However, the resolver is a very important part here. I can imagine that we would like to load some source files differently very soon, e.g., the standard library. Most likely, this would require a different resolver. So it would be good to propagate the resolver that is constructed elsewhere.

**Review by shonfeder — APPROVED** [review]
As per our convo, approving to make clear I’m fine with this being merged to unblock work, my other concerns notwithstanding.

File: quint/src/sourceResolver.ts
Comment (shonfeder) [reply quint/src/sourceResolver.ts:47]: I see. That makes sense. Thanks.

File: quint/src/sourceResolver.ts
Comment (shonfeder) [reply quint/src/sourceResolver.ts]: But why the same computation twice , just once in a thunk? You already have the result, so why ever call `toSourceName`?

File: quint/src/quintParserFrontend.ts
Comment (konnov) [reply quint/src/quintParserFrontend.ts:152]: > sorry if i'm being dense, but i'm afraid I don't follow: you are already copying the Phase1 output into a map via the `new Map` call, no? Each time you get a new Phase1 result you are adding it to this map. Whether you mutate this map or mutate the ref to add a new map, you are not touching the phase1 result in directly either way, are you?

oh, that's what you mean. This is how I am extending this map:

```js
        sourceMap = new Map([...sourceMap, ...parseResult.value.sourceMap])
```

File: quint/src/sourceResolver.ts
Comment (konnov) [reply quint/src/sourceResolver.ts:47]: Actually, I just have hit this problem with the VSCode plugin. @bugarela and I agreed to handle only `file://` atm, but we should check how it is done with URIs in general, postponed for #831.

**Comment by konnov** [comment]
I have addressed most of the comments, also renamed the parsing phases. It took me a bit longer than expected. But imports seem to work as expected now. Very exciting feature 🔥

The only annoying thing is that REPL does not use the current shell directory to load imports. Postponed for #829.

Merging this PR to unblock @bugarela and @thpani on the IBC work.

**Pull request merged**
